File: ucf_sub_catkin_ros/src/sub_teleop/sub_joystick_control/src/dual_stick_control.cpp
#include <ros/ros.h>
#include <geometry_msgs/Wrench.h>
#include <geometry_msgs/Vector3.h>
#include <sensor_msgs/Joy.h>

class ThrustControl {
  ros::NodeHandle nh_;
  ros::Subscriber trans_sub;
  ros::Subscriber rot_sub;
  ros::Publisher twist_pub;
  
  geometry_msgs::Vector3 linear;
  geometry_msgs::Vector3 angular;
  geometry_msgs::Wrench twistMsg;

public:
  ThrustControl()
  {
    trans_sub = nh_.subscribe("/translate/joy", 1, &ThrustControl::translateCb, this);
    rot_sub = nh_.subscribe("/rotate/joy", 1, &ThrustControl::rotateCb, this);
    
    twist_pub = nh_.advertise<geometry_msgs::Wrench>("/desiredThrustWrench", 1000);
  }
  void init()
  {
    ros::Rate rate(50);
    while(ros::ok()) {
      //twistMsg.linear = linear;
      //twistMsg.angular = angular;
      twist_pub.publish(twistMsg);
      ros::spinOnce();
      rate.sleep();
    }

  }

  void translateCb(const sensor_msgs::Joy &msg)
  {
    if(sizeof(msg.axes)/sizeof(float) < 3) {
      ROS_INFO("JOYSTICK ERROR");
      return;
    }
    twistMsg.force.x = msg.axes[1] * 20;
    twistMsg.force.y = msg.axes[0] * 20;
    twistMsg.force.z = msg.axes[2] * 20;
    ROS_INFO("TRANSLATION UPDATED");
  }

  void rotateCb(const sensor_msgs::Joy &msg)
  {
    if(sizeof(msg.axes)/sizeof(float) < 3) {
      ROS_INFO("JOYSTICK ERROR: Not enough axes");
      return;
    }
    twistMsg.torque.x = msg.axes[0] * -2;
    twistMsg.torque.y = msg.axes[1] * 2;
    twistMsg.torque.z = msg.axes[2] * 2;
    ROS_INFO("ROTATION UPDATED");
  }
};

  int main(int argc, char** argv)
  {
    ros::init(argc, argv, "thrust_control");
    ThrustControl tc;
    tc.init();

    return 0;
  }

File: ucf_sub/src/sub_teleop/sub_joystick_control/src/dual_stick_control.py
import rospy
from geometry_msgs.msg import Wrench
from geometry_msgs.msg import Vector3
from sensor_msgs.msg import Joy

class ThrusterControl:

	def __init__(self):
		self.trans_sub = rospy.Subscriber("/translate/joy", Joy, translateCb)
		self.rot_sub = rospy.Subscriber("/rotate/joy", Joy, rotateCb)
		self.twist_pub = rospy.Publisher("/desiredThrustWrench", Wrench, queue_size=1000)
		self.limit_pub = rospy.Publisher("/thruster_limit", Float32, queue_size=1000)
		self.autonomyEnabled = -1
		self.autonomy_sub = rospy.Subscriber("/autonomyWrench", Wrench, republishWrench)
		self.autonomy_pub = rospy.Publisher("/autonomyWrench", Wrench, queue_size=1000)
		self.twistmsg = Wrench()

	def translateCb(self, msg):
		if(len(msg.axes)/len(float) < 3):
			rospy.loginfo("JOYSTICK ERROR")
			return

		twistMsg.force.x = msg.axes[1] * 20
		twistMsg.force.y = msg.axes[0] * 20
		twistMsg.force.z = msg.axes[2] * 20
		thruster_limit = (msg.axes[3]+1)/4
		self.limit_pub.publish(thruster_limit)
		rospy.loginfo("TRANSLATION UPDATED")


	def rotateCb(self, msg):
		if(len(msg.axes)/len(float) < 3):
			rospy.loginfo("JOYSTICK ERROR: Not enough axes")
			return

		twistMsg.torque.x = msg.axes[0] * 20
		twistMsg.torque.y = msg.axes[1] * 20
		twistMsg.torque.z = msg.axes[2] * 20
		self.autonomyEnabled = msg.axes[3]
		rospy.loginfo("ROTATION UPDATED")


	def republishWrench(self, msg):
		if autonomyEnabled > 0:
			self.twist_pub.publish(msg)

if __name__== "__main__":
	tc = ThrusterControl()
	rospy.init_node('ThrusterControl', anonymous=False)
	rate = rospy.Rate(50)

	while not rospy.is_shutdown(:
		tc.twist_pub.publish(tc.twistMsg)
		rate.sleep()